### PR TITLE
Adding option for broadcasting tensors similar to torch WIP

### PR DESCRIPTION
**Pull request by catch-twenty-two**
This adds functionality to broadcast two tensors with potentially different static ranks to a common rank and then preform an operation on them. Before adding more operations via macros, I wanted to make sure this would be helpful to the community since as of now I use it the following way:

```
let a = Tensor::<B, 6>::empty([7, 6, 2, 3, 1, 9], device);
let b = Tensor::<B, 4>::empty([2, 1, 7, 1], device);

let (a, b) = broadcast!(a: Tensor<B, 6>, b: Tensor<B, 4>);

let a_add_b = a.add(b);
```
More functionality can be easily added to use it as so as well:

```
let a = Tensor::<B, 6>::empty([7, 6, 2, 3, 1, 9], device);
let b = Tensor::<B, 4>::empty([2, 1, 7, 1], device);

let a = add_broadcast!(a: Tensor<B, 6>, b: Tensor<B, 4>);
```
Note: Macros were incorporated since the broadcast operation requires knowledge of what size the rank from the resulting tensors will be (which are constants) prior to compile time.

### Checklist

- [ ] Confirmed that `cargo run-checks` command has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

https://github.com/tracel-ai/burn/issues/3344
https://github.com/tracel-ai/burn/issues/1499

### Changes

I've added the logic to correctly shape the tensors via a broadcast function and have operations done on them.

### Testing

Multiple unit tests, and integrated into my own code base.

### CODE REVIEW

**Review by laggui — COMMENTED** [review]
I think having broadcasting for tensors of different rank would address a real gap in usability 🙂

But we should be a thoughtful about the right approach (and I haven't really thought about it yet).

Some general comments on the current macro:
- it only applies left padding to both operands, but broadcasting can be a bit more complex
- the syntax is a little verbose with the type annotations vs type inference
- having a per-op macro like `add_broadcast!` is helpful to apply different broadcasting rules, but can lead to macro explosion; we can probably have an element-wise broadcast and matmul broadcast to cover almost all cases

**Comment by catch-twenty-two** [comment]
Thanks for the review! 

> I think having broadcasting for tensors of different rank would address a real gap in usability 🙂
> 

So, that kinda sounds like a yay, not a nay on usefulness? 😁

> But we should be a thoughtful about the right approach (and I haven't really thought about it yet).
> 
> Some general comments on the current macro:
> 
>  -  it only applies left padding to both operands, but broadcasting can be a bit more complex
> 
Could you give me an quick example? I thought I had all the bases covered here afa the general way pytorch handles broadcasting. If there are some corner cases I'd love to account for them.

https://docs.pytorch.org/docs/stable/notes/broadcasting.html

>    - the syntax is a little verbose with the type annotations vs type inference

Agreed! I couldn't think of a better way to pass the generic constants to the actual function (since as far as I know, this is the only way to do this?) to compare which one is larger (since this has to be done at compile time) and still have it look intuitively like the rest of the burn/rust tensor type syntax 
```
let a: Tensor<B, D, K> = ect..
```

Any ideas would be greatly appreciated, as I know once this syntax is incorporated to the api, you are pretty much stuck with it.

I've since, updated it a bit, but it's still verbose:
```
broadcast!(
    a: Tensor<Backend, RANK_A>,
    b: Tensor<RANK_B>
)
```
>  - having a per-op macro like `add_broadcast!` is helpful to apply different broadcasting rules, but can lead to macro explosion; we can probably have an element-wise broadcast and matmul broadcast to cover almost all cases

Agreed, this is why I only did it for one... I just wanted to show an example of what could be done... Personally I really dislike macros, I feel that they obfuscate the code and make debugging difficult, the less the better! I'll look into incorporating matmul.

**Comment by laggui** [comment]
Sorry for the late response, notification got lost in the recent flood 😅 

> So, that kinda sounds like a yay, not a nay on usefulness? 😁

Yeah I definitely see the usefulness! Just gotta nail the usage/implementation 🙂 

> Could you give me an quick example? I thought I had all the bases covered here afa the general way pytorch handles broadcasting. If there are some corner cases I'd love to account for them.

Actually this kinda ties into the third point regarding elemwise and matmul broadcast rules. For elemwise I believe it's simply left padding, it's matmul that has different rules.

Not entirely sure yet about the best approach, I'd have to think about the best way to tackle the syntax and rules 🤔

**Comment by antimora** [comment]
@catch-twenty-two is it ready for a review?

**Comment by catch-twenty-two** [comment]
> @catch-twenty-two is it ready for a review? 

Hey @antimora,

It is a start and I think it works well,  I have used it all through-out my projects. After talking with @laggui I tried implementing matmul but ran into issues with the requirement of knowing how many dimensions the tensors have prior to doing specific matmul operations on them. (It sounds like this is an issue that has been discussed before at length) I couldn't have done a lot of what I needed for training/image augmentation with out it so it may be better than nothing ATM, but that's up to you all.

**Comment by catch-twenty-two** [comment]
@antimora If you do decide to use it, I still need to integrate the unit tests into the overall code base, so it's not ready to merge yet. But is ready for a review. Thanks!

**Comment by nathanielsimard** [comment]
I think we will support the same notation with dynamic rank instead of macros

**Review by laggui — COMMENTED** [review]
Still unsure if the macro is the right approach.

We could possibly provide something like `tensor.broadcast_with(other)`.

**Comment by catch-twenty-two** [comment]
> I think we will support the same notation with dynamic rank instead of macros

@nathanielsimard This is the best case scenario! But until that happens this was the best alternative to get me unblocked.

If you have some unlreleased code you wanted to share that uses dynamic rank, I could work that into this instead of using macros...

> Still unsure if the macro is the right approach.
> We could possibly provide something like tensor.broadcast_with(other).

@laggui - I couldn't do this without macros since there isn't a concept of dynamic rank yet. Did you have a different idea that I'm not seeing?

**Comment by antimora** [comment]
@catch-twenty-two 

> @laggui - I couldn't do this without macros since there isn't a concept of dynamic rank yet. Did you have a different idea that I'm not seeing?

There is a design document for dyn ranks: https://github.com/tracel-ai/burn/issues/3624 still not implemented. 

I am closing this PR since we've decided not to purse macro direction.
